src/core/shelf.sol

```diff
@@ -80,54 +91,105 @@ contract Shelf is TitleOwned {
+
+        uint loan = title.issue(msg.sender);
+        nftlookup[nft] = loan;
+
+        shelf[loan].registry = registry;
+        shelf[loan].tokenId = token;
+
+        return loan;
     }
 
-    // Move the NFT out of the shelf
-    function move(uint loan, address registry_, uint nft_, address to) public owner(loan) {
-        NFTLike(registry_).transferFrom(address(this), to, nft_);
+    function close(uint loan) public owner(loan) {
+        require(pile.debt(loan) == 0, "outstanding-debt");
+        title.close(loan);
+        bytes32 nft = keccak256(abi.encodePacked(shelf[loan].registry, shelf[loan].tokenId));
+        nftlookup[nft] = 0;
     }
 
-    function release (uint loan, address usr) public owner(loan) {
-        uint debt = pile.debtOf(loan);
-        require(debt == 0, "debt");
-        move(loan, shelf[loan].registry, shelf[loan].tokenId, usr);
-        adjust(loan);
+    // --- Currency actions ---
+    function want() public view returns (int) {
+        return int(balance) - int(tkn.balanceOf(address(this)));
     }
 
-    function free(uint loan, address usr) public auth  {
-        NFTLike(shelf[loan].registry).transferFrom(address(this), usr, shelf[loan].tokenId);
-        adjust(loan);
+    function borrow(uint loan, uint wad) public owner(loan) {
+        require(nftLocked(loan), "nft-not-locked");
+        ceiling.borrow(loan, wad);
+        pile.accrue(loan);
+        pile.incDebt(loan, wad);
+
+        balances[loan] = add(balances[loan], wad);
+        balance = add(balance, wad);
     }
 
-    function deposit (uint loan, address usr) public owner(loan) {
+    function withdraw(uint loan, uint wad, address usr) public owner(loan) note {
+        require(nftLocked(loan), "nft-not-locked");
+        require(wad <= balances[loan], "amount-too-high");
+        balances[loan] = sub(balances[loan], wad);
+        balance = sub(balance, wad);
+        tkn.transferFrom(address(this), usr, wad);
+    }
+
+    function repay(uint loan, uint wad) public owner(loan) note {
+        require(nftLocked(loan), "nft-not-locked");
+        require(balances[loan] == 0,"before repay loan needs to be withdrawn");
+        doRepay(loan, msg.sender, wad);
+    }
+
+    function recover(uint loan, address usr, uint wad) public auth {
+        doRepay(loan, usr, wad);
+        uint loss = pile.debt(loan);
+        pile.decDebt(loan, loss);
+    }
+
+    function doRepay(uint loan, address usr, uint wad) internal {
+        pile.accrue(loan);
+        uint loanDebt = pile.debt(loan);
+        // only repay max loan debt
+        if (wad > loanDebt) {
+            wad = loanDebt;
+        } 
+
+        tkn.transferFrom(usr, address(this), wad);
+        pile.decDebt(loan, wad);
+        tkn.approve(lender, wad);
+    }
+
+    // --- NFT actions ---
+    function lock(uint loan, address usr) public owner(loan) {
         NFTLike(shelf[loan].registry).transferFrom(usr, address(this), shelf[loan].tokenId);
```

Comment:
I think this could be a security issue.

Lets say A and B both are using Tinlake and both approved their NFTs to be taken by the shelf.

A could transfer the NFT to B.
B approves shelf to take the NFT.
**A could afterwards still trigger the lock of the NFT**

what happen if we require loan owner and nft owner must always be the same by using `msg.sender` instead of `usr`

Reply:
Yes, it might need to be `msg.sender`